hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -0,0 +1,260 @@
+    _pretty(sb, 0, compact = true)
+    sb.result()
+  }
+
+  def identBase: String
+  def _asIdent: String = {
+    val sb = new StringBuilder
+    sb.append(identBase)
+    sb.append("_of_")
+    types.foreachBetween { ty =>
+      sb.append(ty.asIdent)
+    } {
+      sb.append("AND")
+    }
+    sb.append("END")
+    sb.result()
+  }
+
+  def codeOrdering(mb: EmitMethodBuilder, so: Array[SortOrder]): CodeOrdering =
+    codeOrdering(mb, this, so)
+
+  def codeOrdering(mb: EmitMethodBuilder, other: PType, so: Array[SortOrder]): CodeOrdering
+
+  def isIsomorphicTo(other: PBaseStruct): Boolean =
+    size == other.size && isCompatibleWith(other)
+
+  def isPrefixOf(other: PBaseStruct): Boolean =
+    size <= other.size && isCompatibleWith(other)
+
+  def isCompatibleWith(other: PBaseStruct): Boolean =
+    fields.zip(other.fields).forall{ case (l, r) => l.typ isOfType r.typ }
+
+  def truncate(newSize: Int): PBaseStruct
+
+  override def unsafeOrdering(): UnsafeOrdering =
+    unsafeOrdering(this)
+
+  override def unsafeOrdering(rightType: PType): UnsafeOrdering = {
+    require(this.isOfType(rightType))
+
+    val right = rightType.asInstanceOf[PBaseStruct]
+    val fieldOrderings: Array[UnsafeOrdering] =
+      types.zip(right.types).map { case (l, r) => l.unsafeOrdering(r)}
+
+    new UnsafeOrdering {
+      def compare(r1: Region, o1: Long, r2: Region, o2: Long): Int = {
+        var i = 0
+        while (i < types.length) {
+          val leftDefined = isFieldDefined(r1, o1, i)
+          val rightDefined = right.isFieldDefined(r2, o2, i)
+
+          if (leftDefined && rightDefined) {
+            val c = fieldOrderings(i).compare(r1, loadField(r1, o1, i), r2, right.loadField(r2, o2, i))
+            if (c != 0)
+              return c
+          } else if (leftDefined != rightDefined) {
+            val c = if (leftDefined) -1 else 1
+            return c
+          }
+          i += 1
+        }
+        0
+      }
+    }
+  }
+
+  def nMissingBytes: Int
+
+  def missingIdx: Array[Int]
+
+  def byteOffsets: Array[Long]
+
+  def allocate(region: Region): Long = {
+    region.allocate(alignment, byteSize)
+  }
+
+  def allocate(region: Code[Region]): Code[Long] = region.allocate(alignment, byteSize)
+
+  def setAllMissing(off: Code[Long]): Code[Unit] = {
+    var c: Code[Unit] = Code._empty
+    var i = 0
+    while (i < nMissingBytes) {
+      c = Code(c, Region.storeByte(off + i.toLong, const(0xFF.toByte)))
+      i += 1
+    }
+    c
+  }
+
+  def clearMissingBits(region: Region, off: Long) {
+    var i = 0
+    while (i < nMissingBytes) {
+      Region.storeByte(off + i, 0.toByte)
+      i += 1
+    }
+  }
+
+  def clearMissingBits(off: Code[Long]): Code[Unit] = {
+    var c: Code[Unit] = Code._empty
+    var i = 0
+    while (i < nMissingBytes) {
+      c = Code(c, Region.storeByte(off + i.toLong, const(0)))
+      i += 1
+    }
+    c
+  }
+
+  def clearMissingBits(region: Code[Region], off: Code[Long]): Code[Unit] =
```

Comment:
Almost all of the methods below should be implemented on the PBaseStruct abstract superclass. Anything that calls another method (as an alias) should be there, with only the ones that load/manipulate memory using Region defined here.

Reply:
Wait, I had cleared this with you: https://hail.zulipchat.com/#narrow/pm-with/116279,130871-pm/near/183208164

Reply:
(and the abstract class had no implementations, in the PArray PR that was PContainer)

Reply:
ah, I shouldn't have answered so broadly there. I think having no implemented methods on the PType interfaces is a too-strong and unnecessary property to enforce. It looks like there are only a couple of places in PArray where we should implement methods in terms of others (isElementMissing / isElementDefined) but there are lots more here.